Add a guide for setting up Access Graph on-prem (#37661)
gravitational/teleport: docs/pages/access-graph/getting-started-on-prem.mdx

---
title: Getting started with Teleport Access Graph (TAG)
---

<Notice type="note">
Teleport Access Graph is provided by default to users of the Teleport Team plan.
Visit the [signup page](https://goteleport.com/signup/) to begin your free&nbsp;trial.

> </Notice>

Author: Wanted to add this call-out, since this "Getting started" guide is the only one we have for now (there's no equivalent for Team/Cloud - but there's nothing to set up manually on Team either).

Reviewer: This almost reads as if the first step to a "getting started on prem" guide is to sign up for a cloud trial.

Can we clarify by saying something like:

"Teleport Access Graph is enabled by default for Teleport Team users. You can skip the setup process here and start using Access Graph right away by signing up for a Teleport Team trial."

## Prerequisites

<Tabs>
<TabItem
  scope={["enterprise"]} label="Self-Hosted Enterprise">

> - A running Teleport Enterprise cluster v15.0.0 or later.

Author: Self-hosted TAG should also work with the next v14 version (v14.3.4). Will update when such version is out.

Reviewer 2: It's out now

Author: Not yet AFAIK? The latest is 14.3.3, pre self-hosted support. https://github.com/gravitational/teleport/releases/tag/v14.3.3

Reviewer 2: It's cut and you can download it from https://goteleport.com/download/
Roman will create the release in github

> - Docker version v(=docker.version=) or later.

Reviewer 3: Technically, docker is not required. Any container runtime will work.

Author: That's true. However, our other guides that use containerized images seem to point towards Docker in a similar way, e.g. https://goteleport.com/docs/management/export-audit-events/fluentd/#prerequisites

Not sure how we'd change up instructions to `docker run`. Maybe somehow say "Docker or other container runtime, but we provide instructions for Docker here" 🙂

Reviewer 3: Let's leave Docker for now. I'd assume that most people understand how to run a container.

> * A PostgresSQL database server v14 or later.

Author: I validated this by running integration tests with containerized Postgres v14. https://github.com/gravitational/access-graph/pull/429 introduces a Postgres 14+ feature (recursive CTEs), so in a very near future we will not be supporting 13 and earlier.

> - A PostgresSQL database server v14 or later.

Reviewer 3: I'd add that RDS is also supported.

Author: Good point. We also have IAM auth support, I can try to document it (+ validate since it has sat unused for a bit).

>   - Access Graph needs a dedicated [database](https://www.postgresql.org/docs/current/sql-createdatabase.html) to store its data.

Reviewer 2: ```suggestion
  - Access Graph needs a dedicated PostgresSQL compatible database to store its data.
```

should we refer as postgresql compatible database?

Author: I'm not so sure we should commit to "PostgreSQL-compatible", since there is no one definition of what that means. E.g. CockroachDB claims to be "compatible", but we have already found that TAG does not currently run on Cockroach.

Author: Btw the list sub-item talks about a "database" as an object inside of Postgres - one Postgres server/cluster can host many "databases", not a database as in an instance.

Reviewer 3: I'd be careful as Cockarouch is Postgres compatible, but TAG won't work with it. I also don't know what other backends we will include with this phrase.

Reviewer 2: I mean, not even cockroachdb say they are compatible with postgres 
> CockroachDB supports the [PostgreSQL wire protocol](https://www.postgresql.org/docs/current/protocol.html) and the majority of PostgreSQL syntax. This means that existing applications built on PostgreSQL can often be migrated to CockroachDB without changing application code.

They implement the wire protocol and some functionalities but they do not claim to be fully compatible

>     The user that TAG connects to Postgres with needs to be an owner of this database, or have similar broad permissions:

Reviewer 2: ```suggestion
    The user that TAG connects to database with needs to be an owner of this database, or have similar broad permissions:
```

    at least the `CREATE TABLE` privilege on the `public` schema, and the `CREATE SCHEMA` privilege.
* A TLS certificate for the Access Graph service
* The node running the Access Graph service must be reachable from Teleport Auth and Proxy services.

> 

Reviewer 3: @klizhentas Good question. Ideas?

> 

Reviewer 4: just add a note section

</TabItem>
</Tabs>

## Step 1/3. Set up the Teleport Access Graph service

> You'll need a copy of your Teleport cluster's host certificate authority (CA) on the machine that hosts the Access Graph service.

Reviewer 2: Can we give a bit of insight why this is necessary? 
Refer that Auth and Proxy will use certificates signed by HostCA to dial to...

---
title: Getting started with Teleport Access Graph (TAG)
---

<Notice type="note">
Teleport Access Graph is provided by default to users of the Teleport Team plan.
Visit the [signup page](https://goteleport.com/signup/) to begin your free&nbsp;trial.
</Notice>

## Prerequisites

<Tabs>
<TabItem
  scope={["enterprise"]} label="Self-Hosted Enterprise">

- A running Teleport Enterprise cluster v14.3.4 or later.
- An updated `license.pem` with Teleport Policy enabled.
- Docker version v(=docker.version=) or later.
- A PostgreSQL database server v14 or later.
  - Access Graph needs a dedicated [database](https://www.postgresql.org/docs/current/sql-createdatabase.html) to store its data.
    The user that TAG connects to the database with needs to be an owner of this database, or have similar broad permissions:
    at least the `CREATE TABLE` privilege on the `public` schema, and the `CREATE SCHEMA` privilege.
  - Amazon RDS for PostgreSQL is supported.

> - A TLS certificate for the Access Graph service

Reviewer: If this is not discussed below, we should clarify the cert requirements (what domains, for example)

TAG requires incoming connections to be authenticated via host certificates that the host CA issues to Auth and Proxy services.

The host CA can be retrieved and saved into a file in one of the following ways:

<Tabs>
<TabItem label="Via curl">
```console
$ mkdir /etc/access_graph
$ curl 'https://<Var name="teleport.example.com" />/webapi/auth/export?type=tls-host' > /etc/access_graph/teleport_host_ca.pem
```
</TabItem>

<TabItem label="Via tctl">
```console
$ mkdir /etc/access_graph
$ tsh login --proxy=<Var name="teleport.example.com" />
$ tctl get cert_authorities --format=json \
    | jq -r '.[] | select(.spec.type == "host") | .spec.active_keys.tls[].cert' \
    | base64 -d > /etc/access_graph/teleport_host_ca.pem
```
</TabItem>
</Tabs>

Then, create a configuration file for the TAG service, similar to this:

```yaml
backend:
  postgres:
    # This uses the PostgreSQL connection URI format, see https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING-URIS
    # A stricter `sslmode` value is strongly recommended,
    # e.g. `sslmode=verify-full&sslrootcert=/etc/access_graph/my_postgres_ca.crt`.

>     # For a full reference on possible parameters see https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS

Reviewer 3: nit: Do you know if our connection string is parsed in any way by `pgx`? If yes, then the supported args could be different.

Author: `pgx` would have to parse it as they don't use `libpq` AFAIK. However, they link to the same documentation, so I'd assume they keep 100% (or close to it) parity https://pkg.go.dev/github.com/jackc/pgx/v5@v5.5.3/pgconn#ParseConfig

    connection: postgres://access_graph_user:mypassword@db.example.com:5432/access_graph_db?sslmode=require

    # When running on Amazon RDS, IAM auth via credentials set in the environment can be used as follows:
    # iam:
    #   aws_region: us-west-2

# IP address (optional) and port for the TAG service to listen to.
# This is the default value. This key can be omitted to listen on port 50051 on all interfaces.
address: ":50051"

tls:
  # File paths of PEM-encoded TLS certificate and private key for the TAG server.
  cert: /etc/access_graph/tls.crt

>   key: /etc/access_graph/tls.key

Reviewer 3: This guide doesn't mention anything about how those certificates should be generated. Here is my proposal https://github.com/gravitational/access-graph/issues/445

# This lists the file paths for host CAs of Teleport clusters that are allowed to register with this TAG service.
# Several paths can be included to allow several Teleport clusters to connect to the TAG service.
registration_cas:
  - /etc/access_graph/teleport_host_ca.pem # A full path to the file containing the Teleport cluster's host CA certificate.
```

Finally, start the TAG service using Docker as follows:

```console
$ docker run -v <path-to-config>:/app/config.yaml -v /etc/access_graph:/etc/access_graph public.ecr.aws/gravitational/access-graph:(=access_graph.version=)
```

## Step 2/3. Update Teleport Auth server configuration

In the YAML config for the Auth service, add a new top-level section for Access Graph configuration.

```yaml
access_graph:
  enabled: true
  # host:port where the TAG service is listening
  endpoint: access-graph.example.com:50051
  # Optionally, specify a trusted CA we expect the TAG service certificate to be signed by.

>   # If not specified, the system trust store will be used.

Reviewer 3: I'd rephrase it. Starting with `optionally` suggests that you can skip this step, which is only true when the TAG's CA is added to a system store.

  ca: /etc/access_graph_ca.pem
```

Then, restart Auth servers, followed by Proxy servers.

## Step 3/3. View the Access Graph in the Web UI

You can find Access Graph in the "Access Management" tab in the Web UI.
![Access Management menu item](../../img/access-graph/menu-item.png)

To access the interface, your user must have a role that allows `list` and `read` verbs on the `access_graph` resource, e.g.:

```yaml
kind: role

> version: v5

Reviewer 3: nit: Do you use v6 or v7 right now by default?

Author: Updated to v6, since that is the default on v14 and v15.

metadata:
  name: my-role
spec:
  allow:
    rules:
    - resources:
      - access_graph
      verbs:
      - list
      - read
```

> The automatically provisioned `editor` role has the required permissions by default.

Reviewer: ```suggestion
The preset `editor` role has the required permissions by default.
```

---

> title: Getting started with Teleport Access Graph (TAG)

Reviewer 5: ```suggestion
title: Run Teleport Access Graph on Self-Hosted Clusters
```
For precision

---

> <Notice type="note">

Reviewer 5: No need for tags here. I would also add a paragraph here that summarizes the scope of this page (what would a user hope to learn by following it?).

Teleport Access Graph is enabled by default for Teleport Team users.
You can skip the setup process here and start using Access Graph right away
by signing up for a [Teleport Team trial](https://goteleport.com/signup/).
</Notice>

## Prerequisites

> <Tabs>

Reviewer 5: No need for Tabs if there's a single TabItem.

  scope={["enterprise"]} label="Self-Hosted Enterprise">

> - A running Teleport Enterprise cluster v14.3.4 or later.

Reviewer 6: ```suggestion
- A running Teleport Enterprise cluster v14.3.4 or later.
- An updated `license.pem` with Teleport Access Graph enabled. 
```

Author: Thanks, I added this to the list (as "Teleport Policy", as that's the product we sell), and removed the standalone notice box.

- A running Teleport Enterprise cluster v14.3.4 or later.
- An updated `license.pem` with Teleport Policy enabled.
- Docker version v(=docker.version=) or later.
- A PostgreSQL database server v14 or later.
  - Access Graph needs a dedicated [database](https://www.postgresql.org/docs/current/sql-createdatabase.html) to store its data.
    The user that TAG connects to the database with needs to be an owner of this database, or have similar broad permissions:
    at least the `CREATE TABLE` privilege on the `public` schema, and the `CREATE SCHEMA` privilege.
  - Amazon RDS for PostgreSQL is supported.
- A TLS certificate for the Access Graph service
  - The TLS certificate must be issued for "server authentication" key usage,
    and must list the IP or DNS name of the TAG service in an X.509 v3 `subjectAltName` extension.

> - The node running the Access Graph service must be reachable from Teleport Auth and Proxy services.

Reviewer 5: This is the first mention of the architecture of the setup. Since architecture often trips up users, we should add a brief description of the components involved. (I'd recommend doing this before the Prerequisites.)

</TabItem>
</Tabs>

## Step 1/3. Set up the Teleport Access Graph service

You'll need a copy of your Teleport cluster's host certificate authority (CA) on the machine that hosts the Access Graph service.

> TAG requires incoming connections to be authenticated via host certificates that the host CA issues to Auth and Proxy services.

Reviewer 5: ```suggestion
TAG requires incoming connections to be authenticated via host certificates that the host CA issues to the Auth Service and Proxy Service.
```
Since these are proper names in the context of Teleport

The host CA can be retrieved and saved into a file in one of the following ways:

<Tabs>
<TabItem label="Via curl">

> ```console

Reviewer 5: ```suggestion
```code
```
For consistency with other docs pages

$ mkdir /etc/access_graph
$ curl 'https://<Var name="teleport.example.com" />/webapi/auth/export?type=tls-host' > /etc/access_graph/teleport_host_ca.pem
```
</TabItem>

<TabItem label="Via tctl">

> ```console

Reviewer 5: ```suggestion
```code
```

$ mkdir /etc/access_graph
$ tsh login --proxy=<Var name="teleport.example.com" />
$ tctl get cert_authorities --format=json \
    | jq -r '.[] | select(.spec.type == "host") | .spec.active_keys.tls[].cert' \
    | base64 -d > /etc/access_graph/teleport_host_ca.pem
```
</TabItem>
</Tabs>

> Then, create a configuration file for the TAG service, similar to this:

Reviewer 5: We should make it more explicit that this configuration file should be on the server running the TAG service

Author: We do mention it at the start of this section https://github.com/gravitational/teleport/pull/37661/files/e9b522fce41ba99443d5151c2e8ff9482b124336#diff-351d93272d6e47e36224548eec736412a333de6d70c937e1512ccfaf61faa8f0R35 , but I'll re-iterate here.

    connection: postgres://access_graph_user:mypassword@db.example.com:5432/access_graph_db?sslmode=require

    # When running on Amazon RDS, IAM auth via credentials set in the environment can be used as follows:
    # iam:
    #   aws_region: us-west-2

# IP address (optional) and port for the TAG service to listen to.
# This is the default value. This key can be omitted to listen on port 50051 on all interfaces.
address: ":50051"

tls:
  # File paths of PEM-encoded TLS certificate and private key for the TAG server.
  cert: /etc/access_graph/tls.crt
  key: /etc/access_graph/tls.key

# This lists the file paths for host CAs of Teleport clusters that are allowed to register with this TAG service.
# Several paths can be included to allow several Teleport clusters to connect to the TAG service.
registration_cas:
  - /etc/access_graph/teleport_host_ca.pem # A full path to the file containing the Teleport cluster's host CA certificate.
```

Finally, start the TAG service using Docker as follows:

```console
$ docker run -v <path-to-config>:/app/config.yaml -v /etc/access_graph:/etc/access_graph public.ecr.aws/gravitational/access-graph:(=access_graph.version=)
```

> ## Step 2/3. Update Teleport Auth server configuration

Reviewer 5: ```suggestion
## Step 2/3. Update the Teleport Auth Service configuration
```
For consistency in product names

> In the YAML config for the Auth service, add a new top-level section for Access Graph configuration.

Reviewer 5: ```suggestion
In the YAML config for the Auth Service, add a new top-level section for Access Graph configuration.
```
Product name consistency

```yaml
access_graph:
  enabled: true
  # host:port where the TAG service is listening
  endpoint: access-graph.example.com:50051
  # Specify a trusted CA we expect the TAG server certificate to be signed by.
  # If not specified, the system trust store will be used.
  ca: /etc/access_graph_ca.pem
```

> Then, restart Auth servers, followed by Proxy servers.

Reviewer 5: ```suggestion
Then, restart Auth Service instances, followed by Proxy Service instances.
```
Product name consistency
